### ISSUE TITLE
Current cove-ocds master does not deploy due to salt's handling of requirements files

### ISSUE THREAD

**Issue by Bjwebb** [open-contracting/deploy#130]
I just updated requirements in cove-ocds with the latest version of pip-tools https://github.com/open-contracting/cove-ocds/pull/32
This adds a bunch of comments along the lines of `# via -r requirements.in, libcove, libcoveocds, libcoveweb`.
Salt does its own processing of requirements includes, which doesn't respect comments. With the comment above, this tries to fetch a file called `requirements.in, libcove, libcoveocds, libcoveweb`, and fails.
This causes the deploy to fail:

```
         ID: /home/cove/cove/.ve/
    Function: virtualenv.managed
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/var/tmp/.root_ddb04c_salt/pyall/salt/state.py", line 1981, in call
                  **cdata['kwargs'])
                File "/var/tmp/.root_ddb04c_salt/pyall/salt/loader.py", line 1977, in wrapper
                  return f(*args, **kwargs)
                File "/var/tmp/.root_ddb04c_salt/pyall/salt/states/virtualenv_mod.py", line 320, in managed
                  **kwargs
                File "/var/tmp/.root_ddb04c_salt/pyall/salt/modules/pip.py", line 680, in install
                  user=user
                File "/var/tmp/.root_ddb04c_salt/pyall/salt/modules/pip.py", line 351, in _process_requirements
                  reqs = _resolve_requirements_chain(requirement)
                File "/var/tmp/.root_ddb04c_salt/pyall/salt/modules/pip.py", line 281, in _resolve_requirements_chain
                  chain.extend(_resolve_requirements_chain(_find_req(req_file)))
                File "/var/tmp/.root_ddb04c_salt/pyall/salt/modules/pip.py", line 281, in _resolve_requirements_chain
                  chain.extend(_resolve_requirements_chain(_find_req(req_file)))
                File "/var/tmp/.root_ddb04c_salt/pyall/salt/modules/pip.py", line 257, in _find_req
                  with salt.utils.files.fopen(link) as fh_link:
                File "/var/tmp/.root_ddb04c_salt/pyall/salt/utils/files.py", line 399, in fopen
                  f_handle = open(*args, **kwargs)  # pylint: disable=resource-leakage
              IOError: [Errno 2] No such file or directory: u'requirements.in, libcove, libcoveocds, libcoveweb'
```

Salt only feels the need to do this processing if a user if set in the virtualenv function.

Therefore, I think the options here are:
* Update the cove-ocds requirements to alter the comments, and add a note to [our requirements docs](https://ocds-standard-development-handbook.readthedocs.io/en/latest/coding/index.html#id1). This is probably the easiest route.
* Don't set a user on the virtualenv function. We might need to chown it seperately.
* Don't use the virtualenv managed state (e.g. run pip directly)

**Comment by odscjames** [comment]
> Update the cove-ocds requirements to alter the comments, and add a note to our requirements docs. This is probably the easiest route.

Those comments could be really helpful tho. I would really like not to lose those. 

Also, we have such a simple requirements upgrade process at the moment, it would  be nice not to make that more complex.

Not having looked at this in depth yet, my instinct is: 

> Don't use the virtualenv managed state (e.g. run pip directly)

Is easiest, most robust, safest and I'd try this first.


Also, should we check if this needs reporting upstream?

**Comment by Bjwebb** [comment]
Here's the upstream issue: https://github.com/saltstack/salt/issues/56514

**Comment by jpmckinney** [comment]
I've deployed Toucan and others that have the same condition (requirements files with comments from pip-compile), but haven't seen this error.

Locally, are you running Python 3 or Python 2 when deploying? The linked issue suggests that people have reproduce on Python 2 but not Python 3.

**Comment by Bjwebb** [comment]
I think the comments were changed in pip-tools 5.0.0, which was released last Friday (April 16th). Older pip-tools don't add the `-r requirements.in,`.

I'll look into Python 2 vs Python 3, but having looked briefly at the salt internals, I suspect the difference is more Linux vs Windows.

**Comment by jpmckinney** [comment]
I think it'd be better to change `requirements_dev.in` files to use `pip-tool<5` with a comment of `# https://github.com/saltstack/salt/issues/56514`, rather than change deployment.

**Comment by odscjames** [comment]
I'm not totally sure compile-tools 5 is the problem here. I just tried downgrading the version of pip-tools to 4.5 and I still got files with `-r requirements.in`.

It's also more complex than this; there is an interaction with versions of pip and versions of compile-tool. If memory is correct, I had to upgrade pip-tools to v5 on one project because I ended up running with a newer version of pip and pip-tools V4 couldn't handle it. So we will have to upgrade to v5 at some point.

I think changing one deploy repo for our specific case is better than changing multiple software repos. The software repos may be used elsewhere so adding complexity to their build processes to solve a salt problem we have may cause issues elsewhere. Is there a reason you don't want to change deployment?

**Comment by Bjwebb** [comment]
Python 2 vs Python 3 for salt-ssh doesn't seem to make a difference.

Before I noticed the most recent comments, I made the required changes to salt: https://github.com/open-contracting/deploy/pull/131. It's in a PR so can be rejected if appropriate.

Sorry, the v5 thing was a bit of an assumption. I can't see anything obviously related in the changelog, and I can't seem to avoid the behaviour by downgrading.

**Comment by jpmckinney** [comment]
I guess we've just been lucky in that no other `requirements.txt` files got the `-r requirements.in` comments until now. PR looks good. I'll just add a comment.

**Comment by jpmckinney** [comment]
Fixed in #131.

> Is there a reason you don't want to change deployment?

When it comes to working around upstream bugs, I prefer the smallest change per repo – e.g. if the bug could have been avoided by constraining the pip-tools version (very small change) or by changing the local Python version (no change).

The reason for my preference is that workarounds have a tendency to live long after upstream is fixed, and the bigger the workaround, the less obvious it is to restore the proper code.

Anyway, in this case, there is no smaller change available, and unless we make commits on top of the changed lines, restoring the code should be as easy as reverting the commit.
